packages/editor/src/components/post-actions/actions.js

```diff
@@ -307,6 +307,35 @@ const trashPostAction = {
 	},
 };
 
+function useTrashPostAction( postType ) {
+	const registry = useRegistry();
+	const { resource, canUserResolvers } = useSelect(
+		( select ) => {
+			const { getPostType, getCachedResolvers } = select( coreStore );
+			return {
+				resource: getPostType( postType )?.rest_base || '',
+				canUserResolvers: getCachedResolvers().canUser,
+			};
+		},
+		[ postType ]
+	);
+	return useMemo(
+		() => ( {
+			...trashPostAction,
+			isEligible( item ) {
+				return (
+					trashPostAction.isEligible( item ) &&
+					registry
+						.select( coreStore )
+						.canUser( 'delete', resource, item.id )
+				);
+			},
+		} ),
+		// eslint-disable-next-line react-hooks/exhaustive-deps
+		[ registry, resource, canUserResolvers ]
```

Comment:
When we do registry.select( coreStore ).canUser( 'delete', resource, item.id ), on the first call it will return false, then the selector starts returning true but nothing is causing a rerender, so the trash action would not be available until a rerender happens e.g: the user closes and reopen the sidebar. We make the memo depend on canUserResolvers as a way to cause a rerender when the resolver finishes. Of course, this is a temporary hacky fix, our actions API needs to offer a long-term solution as we are discussing in https://github.com/WordPress/gutenberg/pull/62647#discussion_r1644336613.